File: release/pypi/src/scythe_sql/preinstalled.py
"""Detects a matching scythe binary already on PATH."""

from __future__ import annotations

import subprocess
from collections.abc import Callable

from scythe_sql.version_utils import extract_version

_ExecFn = Callable[[], str]


def _default_exec() -> str:
    result = subprocess.run(  # noqa: S603
        ["scythe", "--version"],  # noqa: S607
        capture_output=True,
        text=True,
        check=True,
    )
    return result.stdout


def has_matching_path_binary(want_version: str, exec_fn: _ExecFn = _default_exec) -> bool:
    """Checks whether a `scythe` binary already on PATH matches `want_version` exactly.

    Exact equality only: a newer binary on PATH does not satisfy a pin, since
    silently upgrading defeats the point of pinning the dependency version.
    """
    try:
        output = exec_fn()
    except (OSError, subprocess.SubprocessError):
        return False
    return extract_version(output) == want_version

File: release/npm/lib/preinstalled.js
"use strict";

const { execFileSync } = require("node:child_process");
const { extractVersion } = require("./version");

/**
 * Checks whether a `scythe` binary already on PATH matches `wantVersion`
 * exactly. Returns null (never on PATH, or a different version) rather than
 * throwing, since this is a soft short-circuit, not a required condition.
 *
 * Exact equality only: a newer binary on PATH does not satisfy a pin, since
 * silently upgrading defeats the point of pinning the devDependency version.
 *
 * @param {string} wantVersion
 * @param {(cmd: string, args: string[]) => Buffer} [execFn]
 * @returns {boolean}
 */
function hasMatchingPathBinary(wantVersion, execFn = defaultExec) {
  let output;
  try {
    output = execFn("scythe", ["--version"]).toString("utf8");
  } catch {
    return false;
  }
  const found = extractVersion(output);
  return found === wantVersion;
}

function defaultExec(cmd, args) {
  return execFileSync(cmd, args, { stdio: ["ignore", "pipe", "ignore"] });
}

module.exports = { hasMatchingPathBinary };
